src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+        {
+            var mapping = ImmutableDictionary.CreateBuilder<IParameterSymbol, int>();
+            for (var i = 0; i < arguments.Count; i++)
+            {
+                var argumentParameter = semanticFacts.FindParameterForArgument(invocationSemanticModel, arguments[i], cancellationToken);
+                if (argumentParameter is not null)
+                {
+                    mapping.Add(argumentParameter, i);
+                }
+            }
+
+            return mapping.ToImmutable();
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter.
+        /// </summary>
+        public static async Task<Dictionary<TIdentifierNameSyntax, IParameterSymbol>> MapExpressionToParametersAsync(
+            Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is something that should have code actions displayed for it.
+        /// Depends upon the identifiers in the expression mapping back to parameters.
+        /// Does not handle params parameters.
+        /// </summary>
+        private static async Task<bool> ShouldExpressionDisplayCodeActionAsync(
+            Document document, TExpressionSyntax expression,
+            CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IRangeVariableSymbol or ILocalSymbol)
+                {
+                    return false;
+                }
+
+                if (symbol is IParameterSymbol parameter)
+                {
+                    if (parameter.IsParams)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
```

Comment:
this still doesn't feel sufficient.  say i ahve something like:

```var v = [|p * this.PrivateMethod()|];```

i would be able to introduce parameter here, because 'x' would be a parameter. however, i can't inline `this.PrivateMethod()` into the callsites.  Though i could do the overload/trampolines.  

So either we don't support this case entirely, or we only support the subset of transforms that are legal.